presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -136,94 +196,371 @@ public boolean needsInput()
-            lookupSourceProvider = requireNonNull(getFutureValue(lookupSourceProviderFuture));
+        checkState(hasLookupSourceProvider(), "Not ready to handle input yet");
+
+        SpillInfoSnapshot spillInfoSnapshot;
+        try (LookupSourceLease lookupSourceLease = lookupSourceProvider.leaseLookupSource()) {
+            spillInfoSnapshot = SpillInfoSnapshot.from(lookupSourceLease);
+        }
+
+        addInput(page, spillInfoSnapshot);
+    }
+
+    private void addInput(Page page, SpillInfoSnapshot spillInfoSnapshot)
+    {
+        requireNonNull(spillInfoSnapshot, "spillInfoSnapshot is null");
+
+        if (spillInfoSnapshot.hasSpilled) {
+            page = spillAndMaskSpilledPositions(page, spillInfoSnapshot.spillMask);
+            if (page.getPositionCount() == 0) {
+                return;
+            }
         }
 
         // create probe
+        inputPageSpillEpoch = spillInfoSnapshot.spillEpoch;
         probe = joinProbeFactory.createJoinProbe(page);
 
         // initialize to invalid join position to force output code to advance the cursors
         joinPosition = -1;
     }
 
+    private boolean hasLookupSourceProvider()
+    {
+        if (lookupSourceProvider == null) {
+            if (!lookupSourceProviderFuture.isDone()) {
+                return false;
+            }
+            lookupSourceProvider = requireNonNull(getDone(lookupSourceProviderFuture));
+        }
+        return true;
+    }
+
+    private Page spillAndMaskSpilledPositions(Page page, IntPredicate spillMask)
+    {
+        checkState(spillInProgress.isDone(), "Previous spill still in progress");
+        checkNoFailure(spillInProgress);
+
+        if (!spiller.isPresent()) {
+            spiller = Optional.of(partitioningSpillerFactory.create(
+                    probeTypes,
+                    getPartitionGenerator(),
+                    operatorContext.getSpillContext().newLocalSpillContext(),
+                    operatorContext.getSystemMemoryContext().newAggregatedMemoryContext()));
+        }
+
+        PartitioningSpillResult result = spiller.get().partitionAndSpill(page, spillMask);
+        spillInProgress = result.getSpillingFuture();
+        return result.getRetained();
+    }
+
+    public LocalPartitionGenerator getPartitionGenerator()
+    {
+        if (!partitionGenerator.isPresent()) {
+            partitionGenerator = Optional.of(new LocalPartitionGenerator(hashGenerator, lookupSourceFactory.partitions()));
+        }
+        return partitionGenerator.get();
+    }
+
     @Override
     public Page getOutput()
     {
-        if (probe == null && pageBuilder.isEmpty()) {
+        if (!spillInProgress.isDone()) {
+            return null;
+        }
+        checkNoFailure(spillInProgress);
+
+        if (probe == null && pageBuilder.isEmpty() && !finishing) {
             // Fast exit path when lookup source is still being build
```

Comment:
Also isn't the next check (`if (!hasLookupSourceProvider())`) checking for lookup source is still being built?